twilio/jwt/__init__.py

```diff
@@ -140,7 +134,8 @@ def from_jwt(cls, jwt, key=''):
         verify = True if key else False
 
         try:
-            payload = jwt_lib.decode(bytes(jwt), key, options={
+            alg = jwt_lib.get_unverified_header(jwt).get("alg", "HS256")
+            payload = jwt_lib.decode(jwt, key, algorithms=[alg], options={
```

Comment:
This is an unsafe pattern. (Just as unsafe as the code you’re replacing, but still unsafe.) The problem is that if you’re expecting an RS256 token, an attacker should not be able to forge the token without access to the signer’s RSA **private** key; however, if you trust the algorithm specified in the token’s unverified header, then the attacker can substitute an HS256 token hashed with the signer’s RSA **public** key. See:

https://www.rfc-editor.org/rfc/rfc8725.html#section-2.1
https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/

Since Twilio [documents](https://www.twilio.com/docs/iam/access-tokens#jwt-format) that the value of `alg` **must** be `"HS256"`, we should just specify `algorithms=["HS256"]` and not allow the token’s unverified header to modify this expectation.

Reply:
That's a great find — thank you.

Twilio library also uses RS256 in `twilio.jwt.validation.ClientValidationJwt`, which inherits from `twilio.jwt.Jwt`, but upon reading the docs and looking through the code it looks like that class is used to sign HTTP requests to Twilio servers. You upload your public key to Twilio, then use `twilio.http.ValidationClient` to sign requests, as per [the docs](https://www.twilio.com/docs/iam/pkcv/quickstart).

I'm going to change `algorithms` to `["HS256"]` shortly.

Reply:
Looking through the code again, I don't understand why the choice of algorithm is open to overriding if the docs clearly state that it _has_ to be HS256 or RS256 in `t.j.Jwt` or `t.j.v.ClientValidationJwt`, respectively.

Maybe someone from Twilio can chime in? If the choice of algorithm in those cases is an invariant, we should probably make `algorithm` a class variable and not let the user override it?

Reply:
Thanks for the code review @andersk!!

@karls, I'm OK with not letting the user override the choice of algorithm.

Reply:
I've updated the code. It would be great if @thinkingserious or someone else from Twilio could review the changes.

I had some trouble coming up with a way to completely "hide" the choice of algorithm without rewriting a bunch of code. The constructor for `twilio.jwt.Jwt` currently still allows passing in a custom algorithm, but it defaults to `HS256` as before.

`AccessToken`, `ClientCapabilityToken` and `TaskRouterCapabilityToken` are defined to use `HS256` and `ClientValidationJwt` is defined to use `RS256`.

Library users will now be unable to override the algorithm in `Jwt.to_jwt()`. I have also added an additional check in `Jwt.from_jwt()` to make sure that the algorithm in the token header matches the algorithm designated for the particular token. Technically, you can trigger this if you manually construct a `Jwt` (e.g not via one of the Token classes named above, or `ClientValidationJwt`) with, for instance, `HS512` and then decode it with `Jwt.from_jwt()`. I'm not sure how much of an issue this is in practise.

The assumption I've made is the following — a particular type of token should be decoded using the classmethod of its class. E.g if you create a `ClientCapabilityToken`, you should call `ClientCapabilityToken.from_jwt()` to decode it, not `Jwt.from_jwt()`.